design/ux/session_keys_on_ten_gateway.md

```diff
@@ -20,48 +20,45 @@ The TG already manages VKs on behalf of users, so adding SKs is relatively strai
 
 The advantage of the TG managing SKs is that it can return the funds to the EOA anytime.
 
-
 Another advantage is that they can be reused between sessions to avoid unnecessary transactions.
 
 
 ### The flow
 
-The logic of the game UI (javascript) will call an endpoint on the TG that will return a value transfer transaction that the user must sign in their wallet, together with the address of the SK.
+#### Create
+The logic of the game UI (javascript) will call an `/session_key/create` endpoint on the TG, which will return the account address corresponding to the SK. 
+The game UI will create a value transfer transaction from the main account to this address with enough money to prepay for the moves. The user must sign it with their wallet.
+
+The game will call `/session_key/activate/${sk_account}`
+
+After the initialisation, the game will create unsigned move transactions and submit them to the TG. 
+The TG will sign these transactions with the active SK and submit them to the network.
+
+#### Reuse
+
+If the user already has a session key, the game can retrieve it with `/session_key/list`. This will return a list of `address, amount`.
```

Comment:
I've just realised there's a bit of a security issue here. We have an rpc method that returns the user ID, which means if someone makes a webpage with a Connect wallet button then it can:

- request userID
- request list of session keys for that user ID
- submit unsigned transactions to drain their balances

Possible solutions that spring to mind:

1. Remove query method that returns the user's ID, require all communcation to go via MM RPC without leaking the RPC address (not sure if this is feasible with the initial signing part...)

2. Restrict the transactions that a session key can make. For example, when you call /session_key/create you could require a contract address be included in the request body and limit that session_key to creating transactions to that address. This would prevent anyone that knows the userID from draining the funds directly.

Reply:
Tbh I think number 1 needs to be seriously considered because even without session keys this means that whenever you connect your wallet to an app you risk them snooping through your 'private' tx data.

Reply:
Another option is that userIDs could be sessioned. If you make a request to a user ID from a different IP or after a long period of time then you're forced to re-authenticate with your EOA signature...

Reply:
> Tbh I think number 1 needs to be seriously considered because even without session keys this means that whenever you connect your wallet to an app you risk them snooping through your 'private' tx data.

yeah, totally. I was thinking the same